Extended fix OOM Issue #21634 on KerasHub side

praveenhosdrug123 opened the pull request:
## Summary
Applies distributed initialization fix to model presets to resolve OOM errors during initialization of 7B+ parameter models on 8GB TPU devices. This PR updates Keras Hub model backbone and adds a dist_initializer.

## Issue
Token embedding initialization creates large arrays at time of creation, placing all weights on a single device.
Combined with forward passes during backbone initialization, this causes a 2X to 3X memory spike and triggers OOM on TPUs with limited HBM.

## Solution
Updates model backbone (Gemma, Llama, Qwen) to use dist_initializer for token embeddings, sharding weights across TPU devices during instantiation. Validated on 8-device TPU: models that previously OOM'd during backbone initialization now load successfully. 

## Reference
For memory profiling analysis, cache locality theory, validation logs and alternative solutions considered, refer to: [Doc](https://docs.google.com/document/d/1USQj3Tn1uL7iUijbPGLAsE4NiB_BDPlRRGihAgjWj_k)

Related PR: keras-team/keras#21755

## Checklist
- [o] I have added all the necessary unit tests for my change. -- I have added smoke tests since the exact test is difficult to replicate on CPU/GPU runtime. But i have validated the fix using TPUs and the validation logs are inside the document link.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

praveenhosdrug123 commented:
@amitsrivastava78 -- Thanks for the detailed question and the collab example!

The PR fixes the host OOM issue during the build phase.

## Current Failure Point

The model OOMs during the build phase of the ReversibleEmbedding layer, 
specifically during the forward pass of text embedding - before any device 
sharding logic is reached.

**Expected flow:**
build() → add_weight() → assign() → _direct_assign() → jax.device_put(sharded)

**Actual flow with large models:**
Failure happens at build() before the host tensor reaches the sharding logic 
in _direct_assign().

## Why This is Difficult to Reproduce on Single-RAM Systems

Testing on a single large RAM (even when logically divided into shards) 
doesn't reproduce the real memory pressure because:

- The entire tensor still materializes in one physical memory space
- Memory is then distributed logically, but the materialization already happened
- On real multi-device TPUs (8 devices × 8GB), each device has separate physical memory
- The build and forward passes happen at host level, causing real OOM before any 
  device sharding occurs

## Root Cause & Approach

The root cause behind the memory spike is somewhat speculative. I've documented 
the theory and entire approach in detail [here](https://docs.google.com/document/d/1USQj3Tn1uL7iUijbPGLAsE4NiB_BDPlRRGihAgjWj_k). 
Please have a look - if something doesn't sit right with you, happy to discuss 
or investigate further.

## Evidence: TPU Test Results

Error logs from TPU testing show the failure occurs specifically during build 
of the ReversibleEmbedding layer:

File "/.../keras_hub/src/layers/modeling/reversible_embedding.py", line 103, in build
super().build(inputs_shape)

**Before changes:**
[errors_during_forward_pass.txt](https://github.com/user-attachments/files/22983246/errors_during_forward_pass.txt)

**After changes:**
[using_distributed_initializers.txt](https://github.com/user-attachments/files/22983248/using_distributed_initializers.txt)

When using sharding at tensor instantiation, the memory pressure alleviates and the issue resolves.

Happy to run additional experiments or profiling if it would help!

praveenhosdrug123 commented on keras_hub/src/utils/dist_initializer.py:
```diff
@@ -0,0 +1,647 @@
+        mesh_dict.get(axis, 1) for axis in tensor_layout.axes
+    )
+
+    return num_shards_per_axis
```
This is not planned for immediate future use, but i do think it is an important and counter-intuitive concept which should be included for future use. Scaling from a stats POV on distributed frameworks is still rare as a concept and would help future use cases.

praveenhosdrug123 commented:
@hertschuh - Thanks for the feedback and for taking the time to review the document.

I want to clarify the technical issue:
The OOM problem is about large contiguous memory allocation, not total parameter count. Token embeddings are the largest single array and exceed device memory during initialization, even when the full model would fit after sharding.

Thank you for the context on the general solution. A few follow-up questions to help me understand the timeline:

1. What's the expected completion date for the general fix?
2. Will it handle the edge cases mentioned in the document (interleaving, quantization, LoRA)?
3. Will it detect which variables actually need distribution?

The reason I ask: users are blocked on this today for 7B+ models on 8GB TPU devices. 
If the general fix is months out, would it make sense to:
- Merge this targeted fix as a stopgap
- Mark it deprecated once the general solution ships
- Remove it in a future release

Let me know if that's feasible.

amitsrivastava78 commented:
@praveenhosdrug123 thanks for your comments, sorry was held up with other things, i have checked your PR, in principle i do agree with your approach of solving the host OOM by creating sharded tensor on respective device,  however can you please explore the fix to be there only on keras rather than keras hub ? A good starting point(not the full solution) will be to make changes to keras/src/initializers/random_initializers.py and ensure that RandomNormal and RandomUniform are changed so that it can create a sharded tensor on the respective device. This would ensure that solution is truely generic
Let me know if this approach works for you as this will address the concerns raised by @hertschuh  as well.

praveenhosdrug123 commented:
Hi @amitsrivastava78, cc @hertschuh @divyashreepathihalli ,

I understand the preference for a core Keras solution. Let me explain why I chose the routing-based approach in Keras-Hub utils.

Technical reasoning:
1. Domain-specific problem: The OOM issue occurs specifically during Keras-Hub model loading with large embeddings. Keeping the fix scoped to Keras-Hub utils contains it to where the problem actually manifests.
2. Not all sharding is desirable: Adding layout constraints to core RandomNormal/RandomUniform would apply sharding logic universally - but this isn't always correct. Example: Gemma3 has a mixture of text-only and vision components where you explicitly DON'T want to shard embeddings at onset in case its not a text only model. A layout-constraint-only approach would shard regardless, causing incorrect behavior.
3. Controlled opt-in: The distributed initializer approach allows fine-grained control over which variables get sharded during initialization, based on model architecture requirements - not just generic layout rules.

The core issue: We need logic that says "shard THIS variable in THIS model because it's large" - not "shard all variable everywhere."

Next steps:
My understanding is @amitsrivastava78, @hertschuh, and @divyashreepathihalli  should align on what the recommended approach should be. I can see why each perspective has merit, but these are the technical constraints I encountered while working through the problem.

Happy to discuss further, but I'd suggest the team reach consensus on architectural direction first.

Thanks, 
Praveen

hertschuh commented:
> 2. Not all sharding is desirable: Adding layout constraints to core RandomNormal/RandomUniform would apply sharding logic universally - but this isn't always correct. Example: Gemma3 has a mixture of text-only and vision components where you explicitly DON'T want to shard embeddings at onset in case its not a text only model. A layout-constraint-only approach would shard regardless, causing incorrect behavior.
> 3. Controlled opt-in: The distributed initializer approach allows fine-grained control over which variables get sharded during initialization, based on model architecture requirements - not just generic layout rules.

`keras.Variable` knows if there is a sharding, and knows each sharding specifically. So we're not talking about applying random sharding to all variables blindly.

Simply, this is about changing the current sequence:
```
1. call initializer (returns an array on device 0)
2. shard the returned array according to `self._layout`
```
to
```python
if self._layout:
  # initialize and shard at the same time
  value = jax.jit(outsharding=self._layout)(functool.partial(initializer, self.shape, self.dtype)()
else:
  value = initializer(self.shape, self.dtype)
```

Of course this pseudo code is neither complete nor accurate. It require modifying `JaxVariable`, adding overrides, refactoring a bit.

praveenhosdrug123 commented:
Understood. The consensus is to make changes in source Keras only, not KerasHub.

I've made the following changes locally to support sharding token embeddings at initialization:

Core changes in JaxVariable:

Added check_distributed_init()  and overriding the  _initialize_with_initializer method from the Common JaX variable.

check_distributed_init()  validates:

Initializer supports layout parameter
Variable is a token embedding (name == "embeddings" and path == "token_embedding/embeddings")
PartitionSpec contains actual sharding directives
Modified _initialize_with_initializer() to conditionally pass layout to the initializer when distributed init is needed

```python
def _initialize_with_initializer(self, initializer):
    self.set_tensor_layout()
    if self.check_distributed_init(initializer):
        value = self._convert_to_tensor(
            initializer(self._shape, dtype=self._dtype, layout=self._layout)
        )
    else:
        value = self._convert_to_tensor(
            initializer(self._shape, dtype=self._dtype)
        )
    self._initialize(value)
```

These changes specifically target token embedding initializers while keeping everything else constant.

Initialization flow:
JaxVariable (detects & triggers) → Backend-agnostic Random initializers (receives layout) → JAX random (partial function creation) → JAX distribution library (applies sharding)
Refactored relevant components to support this flow. 

Please let me know if this approach doesn't align well with the context of this pull request.

Validation: Token embedding initialization works well for architectures with only token embeddings as its embedding component.

Since this approach is architecture-agnostic, I'm running experiments across different model architectures to validate compositional behavior.  I'll update accordingly if any issues arise.
